README.md
# fluttermoji  <img src="https://user-images.githubusercontent.com/37346450/103101129-1b9a2100-463c-11eb-8a94-b6fbe44bf00f.png" align="right" height="150" />
![Pub Version](https://img.shields.io/pub/v/fluttermoji) [![Generic badge](https://img.shields.io/badge/demo-LIVE-green.svg)](https://psk907.github.io/fluttermoji/) [![GitHub stars](https://img.shields.io/github/stars/psk907/fluttermoji?style=social)](https://github.com/psk907/fluttermoji/stargazers) [![GitHub stars](https://img.shields.io/badge/DM-me-blue?style=flat&logo=telegram)](https://t.me/psk907)

A light-weight and highly customizable SVG graphic set for Flutter, which provides a Customizer Widget, CircleAvatar and other utility functions.

This package provides you three easy-to-use widgets -

| Name | Description | Screenshot | 
|--------|----------|---------- |
|FluttermojiCircleAvatar | Use your fluttermoji anywhere in your Flutter app with a simple customizable widget. Supports material dark theme too.| ![1608830483994](https://user-images.githubusercontent.com/37346450/103071632-009ec100-45ea-11eb-97c4-96c9ec67e204.gif)
|FluttermojiCustomizer | A comprehensize UI to customize the user's fluttermoji. Offers previews of each individual component and whose looks can be tweaked with a `FluttermojiThemeData`|![1608827561239](https://user-images.githubusercontent.com/37346450/154008536-a687828c-dc9d-4a62-aa11-b800d4fb7763.jpg)
|FluttermojiSaveWidget | Renders a save button by default OR can be used as an [InkWell] wrapper for the [child] widget.|![1608827561239](https://user-images.githubusercontent.com/37346450/154008545-8325af7b-58a2-4419-8544-929ffbdbb9ff.jpg)

******
The appearance of the widgets can be tweaked with `FluttermojiThemeData`. It let's you change things like the customizer's background colours, the box decoration etc.
******

Use the given utility functions to send and receive Fluttermoji data from your server/DB efficiently.

| Function Prototype | Description | 
|------------------|---------------|
|String decodeFluttermojifromString(String encodedData)| Decode your string containing the attributes to a SVG and render it by enclosing this string with a SvgPicture.string() | 
| Future\<Map> encodeMySVGtoMap() | Encodes the local user's fluttermoji to a Map denoting the selected attributes |
|Future\<String> encodeMySVGtoString() | Encodes the local user's fluttermoji to a Map denoting the selected attributes , and stringifies the result before returning the value. | 
| Future<List<bool>> clearFluttermoji() | Erases local user's fluttermoji SVG String and selection from local storage |
	
SVG Assets used are derived from [getavataaars.com](https://getavataaars.com/) .

## Screenshots
###  Example app
<a href="https://psk907.github.io/fluttermoji" >Try out the demo on your browser now</a>
<br />
<br />
<img src="https://user-images.githubusercontent.com/37346450/103443014-fd0dd880-4c80-11eb-8955-309bfb66fb4c.jpg" height ="400" />
<img src="https://user-images.githubusercontent.com/37346450/103443015-fed79c00-4c80-11eb-8219-5edab76c9f0f.jpg" height="400" />
<img src="https://user-images.githubusercontent.com/37346450/103443018-01d28c80-4c81-11eb-8336-ebc19de61220.jpg" height="400" />

**Use them in your games or social media apps**

<a href="https://play.google.com/store/apps/details?id=com.statefullyfidgeting.tugofwar" ><img src="https://user-images.githubusercontent.com/37346450/103443017-00a15f80-4c81-11eb-8223-3404a35079aa.jpg" height="400" /></a>

## Usage Instructions
1. Depend on it by importing your package in the ```pubspec.yaml```  file.
	```yaml
	dependencies:
          fluttermoji: latest_version
	```
2. Add the following import to your .dart file
	```dart
	import 'package:fluttermoji/fluttermoji.dart';
	```
3. Add the FluttermojiCircleAvatar widget to display your Fluttermoji where needed.
	```dart
	FluttermojiCircleAvatar();
	```
4. To allow your users to personalize their Fluttermoji, add the following widget and pair it with the above one in your page.
	```dart
	FluttermojiCustomizer();
	```

That's all it takes, simple right ? The two widgets communicate with each other and update in real-time throughout your widget tree.

##  Usage Guidelines
The package offers a ton of features in the simplest way possible, however there are some points worth noting.

- `FluttermojiCircleAvatar` only renders the local user's avatar. To display the avatar of other users', you'll have to use the `SvgPicture.string()` method from `flutter_svg` .
- `FluttermojiCircleAvatar` would render a default avatar until customized and saved by the local user for the first time.
- Use the `canvaskit` renderer when building for web, the default `html` renderer will not work with SVGs.
- The fluttermoji's attributes are saved to local app/browser storage. Clearing app/browser data would mean clearing these attributes as well.
- You can create a `FluttermojiThemeData` instance to configure the look and feel of the widgets to your liking.

## Attributions

- SVG assets from [Fang-Pen Lin](getAvataars.com)'s  [GitHub repository](https://github.com/fangpenlin/avataaars-generator)
- Icons made by <a href="https://www.flaticon.com/authors/freepik" title="Freepik">Freepik</a> from <a href="https://www.flaticon.com/" title="Flaticon">www.flaticon.com</a>
- Animated [GIF](https://user-images.githubusercontent.com/37346450/103071632-009ec100-45ea-11eb-97c4-96c9ec67e204.gif) designed by [Reesha Shenoy](https://github.com/reeshaa) 

## Community
If you find any issues or have some feedback, please raise the same on the GitHub repo or contact me directly.

Do leave a thumbs up if you liked it.


**Happy Fluttering ; )**


CHANGELOG.md
## [1.0.2] - 26/03/2023
 * Upgrades dependencies

## [1.0.1] - 25/10/2022
 * Minor bug fix [#22](https://github.com/psk907/fluttermoji/pull/22#issue-1250729612)

## [1.0.0] - 15/02/2022
 * Adds `FluttermojiThemeData` and `FluttermojiSaveWidget` to the library.
 * Fixes some visual glithes in the appbar of the `FluttermojiCustomizer`.
 * Minor internal refactoring.
 * Updates dependency versions
 * BREAKING CHANGES 
   *  The customizer widget no longer has the top row with "Customize" and the save button, and the `showSaveWidget` property has been removed.
   *  The widget's do not set to Material light/dark system theme automatically anymore. This must be implemented using the [theme] property.
   *  Refer the doc comments for more details.

*****

## [0.2.3] - 03/12/2021

 * Updates dependency versions

## [0.2.2] - 11/10/2021

 * Adds toggle to show/hide save button widget in `FluttermojiCustomizer` widget.
  
## [0.2.1] - 03/05/2021

 * Fixes bug where preview and avatars don't revert to saved version after unsaved edits
 * Adds `clearFluttermoji()` to FluttermojiFunctions
 * Some optimizations

## [0.2.0] - 08/04/2021

 * Updates dependencies to their null-safe stable version
 * Migrates package to Null safety
 * Removes dead code

## [0.2.0-nullsafety.0] - 10/03/2021

 * Updated dependencies to their null-safe version
 * Migrated package to Null safety
 * Remove dead code

## [0.1.2+1] - 30/01/2021

 * (Hotfix) Cleared some lints for a cleaner code

## [0.1.2] - 19/01/2021

 * Fixed a bug in decodeFluttermojifromString()

## [0.1.1] - 01/01/2021

 * Added Material Dark theme support to FluttermojiCircleAvatar
 * Added screenshots to the README.md

## [0.1.0] - 24/12/2020

* Added documentation to many APIs and to the README.md as well.
* Created new utility functions to allow sharing of Fluttermojis to server/DB and decoding them for render using flutter_svg package.
* Fixed the bug on loading Fluttermoji on first launch.
* Fixed standalone customizer issues.
* Description updates
* Other minor changes and fixes.


## [0.0.1] - 23/12/2020

* Initial Release - documentation and instructions pending


lib/fluttermojiController.dart
import 'dart:convert';
import './fluttermoji_assets/style.dart';
import 'package:get/get.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'fluttermojiFunctions.dart';
import 'fluttermoji_assets/fluttermojimodel.dart';
import 'fluttermoji_assets/clothes/clothes.dart';
import 'fluttermoji_assets/face/eyebrow/eyebrow.dart';
import 'fluttermoji_assets/face/eyes/eyes.dart';
import 'fluttermoji_assets/face/mouth/mouth.dart';
import 'fluttermoji_assets/face/nose/nose.dart';
import 'fluttermoji_assets/skin.dart';
import 'fluttermoji_assets/top/accessories/accessories.dart';
import 'fluttermoji_assets/top/facialHair/facialHair.dart';
import 'fluttermoji_assets/top/hairStyles/hairStyle.dart';

/// Brains of the Fluttermoji package
///
/// Built using the getX architecture to allow the two widgets to communicate with each other
///
/// Exposes certain static functions for use by the developer
class FluttermojiController extends GetxController {
  var fluttermoji = "".obs;

  /// Stores the option selected by the user for each attribute
  /// where the key represents the Attribute
  /// and the value represents the index of the selected option.
  ///
  /// Eg: selectedIndexes["eyes"] gives the index of
  /// the kind of eyes picked by the user
  Map<String?, dynamic> selectedOptions = <String?, dynamic>{};

  @override
  void onInit() {
    // called immediately after the widget is allocated memory
    init();
    super.onInit();
  }

  void init() async {
    Map<String?, int> _tempIndexes = await getFluttermojiOptions();
    selectedOptions = _tempIndexes;
    update();
    fluttermoji.value = getFluttermojiFromOptions();
    update();
  }

  /// Adds fluttermoji new string to fluttermoji in GetX Controller
  void updatePreview({
    String fluttermojiNew = '',
  }) {
    if (fluttermojiNew.isEmpty) {
      fluttermojiNew = getFluttermojiFromOptions();
    }
    fluttermoji.value = fluttermojiNew;
    update();
  }

  /// Restore controller state
  /// with the latest SAVED version of [fluttermoji] and [selectedOptions]
  void restoreState() async {
    SharedPreferences pref = await SharedPreferences.getInstance();

    // Replace observable [fluttermoji] with latest saved version or use default attributes if null
    fluttermoji.value = pref.getString('fluttermoji') ??
        FluttermojiFunctions().decodeFluttermojifromString(
          jsonEncode(defaultFluttermojiOptions),
        );

    selectedOptions = await getFluttermojiOptions();
    update();
  }

  String _getFluttermojiProperty(String type) {
    return fluttermojiProperties[type]!
        .property!
        .elementAt(selectedOptions[type] as int);
  }

  ///  Accepts a String [fluttermoji]
  ///
  ///  stores [fluttermoji] in device storage
  ///  adds the new name to controller
  ///
  ///  Thereby updating all the states which are listening to controller
  Future<void> setFluttermoji({String fluttermojiNew = ''}) async {
    if (fluttermojiNew.isEmpty) {
      fluttermojiNew = getFluttermojiFromOptions();
    }
    SharedPreferences pref = await SharedPreferences.getInstance();
    await pref.setString('fluttermoji', fluttermojiNew);
    fluttermoji.value = fluttermojiNew;
    await pref.setString(
        'fluttermojiSelectedOptions', jsonEncode(selectedOptions));
    update();
  }

  /// Generates a [String] fluttermoji from [selectedOptions] pref
  String getFluttermojiFromOptions() {
    String _fluttermojiStyle =
        fluttermojiStyle[_getFluttermojiProperty('style')]!;
    String _clothe = Clothes.generateClothes(
        clotheType: _getFluttermojiProperty('clotheType'),
        clColor: _getFluttermojiProperty('clotheColor'))!;
    String _facialhair = FacialHair.generateFacialHair(
        facialHairType: _getFluttermojiProperty('facialHairType'),
        fhColor: _getFluttermojiProperty('facialHairColor'))!;
    String _mouth = mouth['${_getFluttermojiProperty('mouthType')}'];
    String _nose = nose['Default'];
    String _eyes = eyes['${_getFluttermojiProperty('eyeType')}'];
    String _eyebrows = eyebrow['${_getFluttermojiProperty('eyebrowType')}'];
    String _accessory = accessories[_getFluttermojiProperty('accessoriesType')];
    String _hair = HairStyle.generateHairStyle(
        hairType: _getFluttermojiProperty('topType'),
        hColor: _getFluttermojiProperty('hairColor'))!;
    String _skin = skin[_getFluttermojiProperty('skinColor')];
    String _completeSVG = '''
<svg width="264px" height="280px" viewBox="0 0 264 280" version="1.1"
xmlns="http://www.w3.org/2000/svg"
xmlns:xlink="http://www.w3.org/1999/xlink">
<desc>Fluttermoji on pub.dev</desc>
<defs>
<circle id="path-1" cx="120" cy="120" r="120"></circle>
<path d="M12,160 C12,226.27417 65.72583,280 132,280 C198.27417,280 252,226.27417 252,160 L264,160 L264,-1.42108547e-14 L-3.19744231e-14,-1.42108547e-14 L-3.19744231e-14,160 L12,160 Z" id="path-3"></path>
<path d="M124,144.610951 L124,163 L128,163 L128,163 C167.764502,163 200,195.235498 200,235 L200,244 L0,244 L0,235 C-4.86974701e-15,195.235498 32.235498,163 72,163 L72,163 L76,163 L76,144.610951 C58.7626345,136.422372 46.3722246,119.687011 44.3051388,99.8812385 C38.4803105,99.0577866 34,94.0521096 34,88 L34,74 C34,68.0540074 38.3245733,63.1180731 44,62.1659169 L44,56 L44,56 C44,25.072054 69.072054,5.68137151e-15 100,0 L100,0 L100,0 C130.927946,-5.68137151e-15 156,25.072054 156,56 L156,62.1659169 C161.675427,63.1180731 166,68.0540074 166,74 L166,88 C166,94.0521096 161.51969,99.0577866 155.694861,99.8812385 C153.627775,119.687011 141.237365,136.422372 124,144.610951 Z" id="path-5"></path>
</defs>
<g id="Fluttermoji" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
<g transform="translate(-825.000000, -1100.000000)" id="Fluttermoji/Circle">
<g transform="translate(825.000000, 1100.000000)">''' +
        _fluttermojiStyle +
        '''
<g id="Mask"></g>
<g id="Fluttermoji" stroke-width="1" fill-rule="evenodd">
<g id="Body" transform="translate(32.000000, 36.000000)">

<mask id="mask-6" fill="white">
<use xlink:href="#path-5"></use>
</mask>
<use fill="#D0C6AC" xlink:href="#path-5"></use>''' +
        _skin +
        '''<path d="M156,79 L156,102 C156,132.927946 130.927946,158 100,158 C69.072054,158 44,132.927946 44,102 L44,79 L44,94 C44,124.927946 69.072054,150 100,150 C130.927946,150 156,124.927946 156,94 L156,79 Z" id="Neck-Shadow" opacity="0.100000001" fill="#000000" mask="url(#mask-6)"></path></g>''' +
        _clothe +
        '''<g id="Face" transform="translate(76.000000, 82.000000)" fill="#000000">''' +
        _mouth +
        _facialhair +
        _nose +
        _eyes +
        _eyebrows +
        _accessory +
        '''</g>''' +
        _hair +
        '''</g></g></g></g></svg>''';
    return _completeSVG;
  }

  Future<Map<String?, int>> getFluttermojiOptions() async {
    SharedPreferences pref = await SharedPreferences.getInstance();
    String? _fluttermojiOptions = pref.getString('fluttermojiSelectedOptions');
    if (_fluttermojiOptions == null || _fluttermojiOptions == '') {
      Map<String?, int> _fluttermojiOptionsMap =
          Map.from(defaultFluttermojiOptions);
      await pref.setString(
          'fluttermojiSelectedOptions', jsonEncode(_fluttermojiOptionsMap));
      selectedOptions = _fluttermojiOptionsMap;

      update();
      return _fluttermojiOptionsMap;
    }
    selectedOptions = Map.from(jsonDecode(_fluttermojiOptions));
    update();
    return Map.from(jsonDecode(_fluttermojiOptions));
  }

  String? getComponentTitle(String attributeKey, int attriibuteValueIndex) {
    return fluttermojiProperties[attributeKey]!
        .property
        ?.elementAt(attriibuteValueIndex);
  }

  /// Generates compnonent SVG string for an individual component
  /// to display as a preview
  String getComponentSVG(String? attributeKey, int? attributeValueIndex) {
    switch (attributeKey) {
      case 'clotheType':
        return '''<svg width="100px" height="120px" viewBox="30 100 200 250" >''' +
            Clothes.generateClothes(
                clotheType: ClotheType.elementAt(attributeValueIndex!),
                clColor: ClotheColor[selectedOptions['clotheColor']])! +
            '''</svg>''';

      case 'clotheColor':
        return '''<svg width="120px" height="120px" > 
                <circle cx="60" cy="60" r="35" stroke="black" stroke-width="1" fill="''' +
            Clothes.clotheColor[ClotheColor[attributeValueIndex!]] +
            '''"/></svg>''';

      case 'topType':
        if (attributeValueIndex == 0) return emptySVGIcon;
        return '''<svg width="20px" width="100px" height="100px" viewBox="10 0 250 250">''' +
            HairStyle.generateHairStyle(
                hairType: TopType[attributeValueIndex!],
                hColor: HairColor[selectedOptions['hairColor']])! +
            '''</svg>''';

      case 'hairColor':
        return '''<svg width="120px" height="120px" > 
                <circle cx="60" cy="60" r="30" stroke="black" stroke-width="1" fill="''' +
            HairStyle.hairColor[HairColor.elementAt(attributeValueIndex!)] +
            '''"/> </svg>''';

      case 'facialHairType':
        if (attributeValueIndex == 0) return emptySVGIcon;
        return '''<svg width="20px" height="20px" viewBox="0 -40 112 180" >''' +
            FacialHair.generateFacialHair(
                facialHairType: FacialHairType[attributeValueIndex!],
                fhColor: FacialHairColor[selectedOptions['facialHairColor']])! +
            '''</svg>''';

      case 'facialHairColor':
        return '''<svg width="120px" height="120px" > 
                <circle cx="60" cy="60" r="30" stroke="black" stroke-width="1" fill="''' +
            FacialHair.facialHairColor[FacialHairColor[attributeValueIndex!]] +
            '''"/></svg>''';

      case 'eyeType':
        return '''<svg width="20px" height="20px" viewBox="-3 -30 120 120">''' +
            eyes[EyeType[attributeValueIndex!]] +
            '''</svg>''';

      case 'eyebrowType':
        return '''<svg width="20px" height="20px" viewBox="-3 -50 120 120">''' +
            eyebrow[EyebrowType[attributeValueIndex!]] +
            '''</svg>''';

      case 'mouthType':
        return '''<svg width="20px" height="20px" viewBox="0 10 120 120">''' +
            mouth[MouthType[attributeValueIndex!]] +
            '''</svg>''';

      case 'accessoriesType':
        if (attributeValueIndex == 0) return emptySVGIcon;
        return '''<svg width="20px" height="20px" viewBox="-3 -50 120 170" >''' +
            accessories[AccessoriesType[attributeValueIndex!]] +
            '''</svg>''';

      case 'skinColor':
        return '''<svg width="264px" height="280px" viewBox="0 0 264 280" version="1.1"
xmlns="http://www.w3.org/2000/svg"
xmlns:xlink="http://www.w3.org/1999/xlink">
<desc>Fluttermoji Skin Preview</desc>
<defs>
<circle id="path-1" cx="120" cy="120" r="120"></circle>
<path d="M12,160 C12,226.27417 65.72583,280 132,280 C198.27417,280 252,226.27417 252,160 L264,160 L264,-1.42108547e-14 L-3.19744231e-14,-1.42108547e-14 L-3.19744231e-14,160 L12,160 Z" id="path-3"></path>
<path d="M124,144.610951 L124,163 L128,163 L128,163 C167.764502,163 200,195.235498 200,235 L200,244 L0,244 L0,235 C-4.86974701e-15,195.235498 32.235498,163 72,163 L72,163 L76,163 L76,144.610951 C58.7626345,136.422372 46.3722246,119.687011 44.3051388,99.8812385 C38.4803105,99.0577866 34,94.0521096 34,88 L34,74 C34,68.0540074 38.3245733,63.1180731 44,62.1659169 L44,56 L44,56 C44,25.072054 69.072054,5.68137151e-15 100,0 L100,0 L100,0 C130.927946,-5.68137151e-15 156,25.072054 156,56 L156,62.1659169 C161.675427,63.1180731 166,68.0540074 166,74 L166,88 C166,94.0521096 161.51969,99.0577866 155.694861,99.8812385 C153.627775,119.687011 141.237365,136.422372 124,144.610951 Z" id="path-5"></path>
</defs>
	<g id="Fluttermoji" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
		<g transform="translate(-825.000000, -1100.000000)" id="Fluttermoji/Circle">
			<g transform="translate(825.000000, 1100.000000)">
				<g id="Mask"></g>
        <g id="Fluttermoji" stroke-width="1" fill-rule="evenodd">
					<g id="Body" transform="translate(32.000000, 36.000000)">
						<mask id="mask-6" fill="white">
							<use xlink:href="#path-5"></use>
						</mask>
						<use fill="#D0C6AC" xlink:href="#path-5"></use>
        ''' +
            skin[SkinColor[attributeValueIndex!]] +
            '''	<path d="M156,79 L156,102 C156,132.927946 130.927946,158 100,158 C69.072054,158 44,132.927946 44,102 L44,79 L44,94 C44,124.927946 69.072054,150 100,150 C130.927946,150 156,124.927946 156,94 L156,79 Z" id="Neck-Shadow" opacity="0.100000001" fill="#000000" mask="url(#mask-6)"></path>
				</g>
		</g>
	</g>
</svg>''';

      case 'style':
        return '''<svg width="264px" height="280px" viewBox="0 0 264 280" version="1.1"
xmlns="http://www.w3.org/2000/svg"
xmlns:xlink="http://www.w3.org/1999/xlink">
<desc>Fluttermoji Skin Preview</desc>
<defs>
<circle id="path-1" cx="120" cy="120" r="120"></circle>
<path d="M12,160 C12,226.27417 65.72583,280 132,280 C198.27417,280 252,226.27417 252,160 L264,160 L264,-1.42108547e-14 L-3.19744231e-14,-1.42108547e-14 L-3.19744231e-14,160 L12,160 Z" id="path-3"></path>
<path d="M124,144.610951 L124,163 L128,163 L128,163 C167.764502,163 200,195.235498 200,235 L200,244 L0,244 L0,235 C-4.86974701e-15,195.235498 32.235498,163 72,163 L72,163 L76,163 L76,144.610951 C58.7626345,136.422372 46.3722246,119.687011 44.3051388,99.8812385 C38.4803105,99.0577866 34,94.0521096 34,88 L34,74 C34,68.0540074 38.3245733,63.1180731 44,62.1659169 L44,56 L44,56 C44,25.072054 69.072054,5.68137151e-15 100,0 L100,0 L100,0 C130.927946,-5.68137151e-15 156,25.072054 156,56 L156,62.1659169 C161.675427,63.1180731 166,68.0540074 166,74 L166,88 C166,94.0521096 161.51969,99.0577866 155.694861,99.8812385 C153.627775,119.687011 141.237365,136.422372 124,144.610951 Z" id="path-5"></path>
</defs>
	<g id="Fluttermoji" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
    <g transform="translate(-825.000000, -1100.000000)" id="Fluttermoji/Circle">
			<g transform="translate(825.000000, 1100.000000)">''' +
            fluttermojiStyle[FluttermojiStyle[attributeValueIndex!]]! +
            '''<g id="Mask"></g>
        <g id="Fluttermoji" stroke-width="1" fill-rule="evenodd">
					<g id="Body" transform="translate(32.000000, 36.000000)">
						<mask id="mask-6" fill="white">
							<use xlink:href="#path-5"></use>
						</mask>
						<use fill="#D0C6AC" xlink:href="#path-5"></use>
        ''' +
            skin[SkinColor[1]] +
            '''	<path d="M156,79 L156,102 C156,132.927946 130.927946,158 100,158 C69.072054,158 44,132.927946 44,102 L44,79 L44,94 C44,124.927946 69.072054,150 100,150 C130.927946,150 156,124.927946 156,94 L156,79 Z" id="Neck-Shadow" opacity="0.100000001" fill="#000000" mask="url(#mask-6)"></path>
				</g>
		</g>
	</g>
</svg>''';

      default:
        return emptySVGIcon;
    }
  }

  @Deprecated(
      'No longer used by the library, please use the field `selectedOptions` instead.')
  Map<String?, dynamic> selectedIndexes = <String?, dynamic>{};
}
